dhall-csv/src/Dhall/CsvToDhall.hs

```diff
@@ -1,27 +1,235 @@
+            <> O.help "The first value with the matching type (successfully parsed all the way down the tree) is accepted, even if not the only possible match. (DEFAULT)"
+            )
+    uNone   =  O.flag' UNone
+            (  O.long "unions-none"
+            <> O.help "Unions not allowed"
+            )
+    uStrict =  O.flag' UStrict
+            (  O.long "unions-strict"
+            <> O.help "Error if more than one union values match the type (and parse successfully)"
+            )
+
+type ExprX = Expr Src Void
+
+-- | Parse schema code and resolve imports
+resolveSchemaExpr :: Text  -- ^ type code (schema)
+                  -> IO ExprX
+resolveSchemaExpr code = do
+    parsedExpression <-
+      case Dhall.Parser.exprFromText "\n\ESC[1;31mSCHEMA\ESC[0m" code of
+        Left  err              -> throwIO err
+        Right parsedExpression -> return parsedExpression
+    Dhall.Import.load parsedExpression
+
+typeCheckSchemaExpr :: (Exception e, MonadCatch m)
+                    => (CompileError -> e) -> ExprX -> m ExprX
+typeCheckSchemaExpr compileException expr =
+  case TypeCheck.typeOf expr of -- check if the expression has type
+    Left  err -> throwM . compileException $ TypeError err
+    Right t   -> case t of -- check if the expression has type Type
+      Core.Const Core.Type -> return expr
+      _              -> throwM . compileException $ BadDhallType t expr
 
-dhallFromCsv :: [Data.Csv.NamedRecord] -> Expr Src Void
-dhallFromCsv csv = Core.ListLit mType $ Sequence.fromList $ map convertRecord csv
+
+dhallFromCsv :: Conversion -> ExprX -> [Data.Csv.NamedRecord] -> Either CompileError ExprX
+dhallFromCsv Conversion{..} typeExpr = listConvert (Core.normalize typeExpr)
   where
-    mType :: Maybe (Expr Src Void)
-    mType = case csv of
-        [] -> Just $ Core.App Core.List $ Core.Record $ Dhall.Map.fromList []
-        _ -> Nothing
-    convertRecord :: Data.Csv.NamedRecord -> Expr Src Void
-    convertRecord recordCsv = Core.RecordLit dhallMap
-      where
-        dhallMap = Dhall.Map.fromList $ map (bimap decodeUtf8 convertField) $ HashMap.toList recordCsv
-    convertField :: Data.Csv.Field -> RecordField Src Void
-    convertField = Core.makeRecordField . Core.TextLit . (Core.Chunks []) . decodeUtf8
+    listConvert :: ExprX -> [Data.Csv.NamedRecord] -> Either CompileError ExprX
+    listConvert (Core.App Core.List recordType@(Core.Record _)) [] = return $ Core.ListLit (Just recordType) Sequence.empty
+    listConvert (Core.App Core.List recordType) [] = Left $ Unsupported recordType
+    listConvert (Core.App Core.List recordType) csv = do
+        a <- traverse (recordConvert recordType) csv
+        return $ Core.ListLit Nothing $ Sequence.fromList a
+    listConvert e _ = Left $ Unsupported e
+
+    recordConvert :: ExprX -> Data.Csv.NamedRecord -> Either CompileError ExprX
+    recordConvert (Core.Record record) csvRecord
+        | extraKeys <- (map decodeUtf8 $ HashMap.keys csvRecord) \\ Map.keys record
```

Comment:
I recommend using `decodeUtf8'` and handling the error, even if just to explain in the error message that the CSV is malformed

Same comment for other uses of `decodeUtf8`

Reply:
I changed it, in this case I'm still using `decodeUtf8` to keep it simple but I check in a previous guard with `decodeUtf8` if there are any errors. LMK if that's fine